src/apps/service/oauth.clj

```diff
@@ -69,15 +69,25 @@
 (defn authorization-uri
   "Generates an authorization URI for a remote API."
   [{:keys [api-name] :as server-info} username state-info]
-  (when-not (op/get-access-token api-name username)
-    (str (assoc (curl/url (:auth-uri server-info))
-                :query {:response_type "code"
-                        :client_id     (:client-key server-info)
-                        :redirect_uri  (:redirect-uri server-info)
-                        :state         (op/store-authorization-request username state-info)}))))
+  (str (assoc (curl/url (:auth-uri server-info))
+              :query {:response_type "code"
+                      :client_id     (:client-key server-info)
+                      :redirect_uri  (:redirect-uri server-info)
+                      :state         (op/store-authorization-request username state-info)})))
+
+(defn has-access-token
+  "Determines whether or not a user has an access token for an external API."
+  [{:keys [api-name] :as server-info} username]
+  (seq (op/get-access-token api-name username)))
+
+(defn- build-auth-uri
+  "Builds an authorization URI for an external API if the user doesn't have an access token already."
+  [server-info username]
+  (when-not (has-access-token server-info username)
```

Comment:
It's not really new, just refactored a little bit. The call to `has-access-token` serves the same purpose as the call to `op/get-access-token` did above. I just moved it out of `authorization-uri`.